fnet/data/bufferedpatchdataset.py

```diff
@@ -1,110 +1,182 @@
+ArrayLike = Union[np.ndarray, torch.Tensor]
+
 
+class BufferedPatchDataset:
+    """Provides patches from items of a dataset.
 
-class BufferedPatchDataset(FnetDataset):
-    """Dataset that provides chunks/patchs from another dataset."""
+    Parameters
+    ----------
+    dataset
+        Dataset object.
+    patch_shape
+        Shape of patch to be extracted from dataset items.
+    buffer_size
+        Size of buffer.
+    buffer_switch_interval
+        Number of patches provided between buffer item exchanges. Set to -1 to
+        disable exchanges.
+    shuffle_images
+        Set to randomize order of dataset item insertion into buffer.
+
+    """
 
     def __init__(
-        self,
-        dataset,
-        patch_size=(32, 64, 64),
-        buffer_size=1,
-        buffer_switch_frequency=-1,
-        npatches=100000,
-        verbose=False,
-        shuffle_images=True,
-        dim_squeeze=None,
+            self,
+            dataset: collections.abc.Sequence,
+            patch_shape: Sequence[int] = (32, 64, 64),
+            buffer_size: int = 1,
+            buffer_switch_interval: int = -1,
+            shuffle_images: bool = True,
     ):
-        self.counter = 0
         self.dataset = dataset
+        self.patch_shape = patch_shape
         self.buffer_size = min(len(self.dataset), buffer_size)
-        self.buffer_switch_frequency = buffer_switch_frequency
-        self.npatches = npatches
-        self.buffer = list()
-        self.verbose = verbose
+        self.buffer_switch_interval = buffer_switch_interval
         self.shuffle_images = shuffle_images
-        self.dim_squeeze = dim_squeeze
-        shuffed_data_order = np.arange(0, len(dataset))
-        if self.shuffle_images:
-            np.random.shuffle(shuffed_data_order)
-        pbar = tqdm(range(0, self.buffer_size))
-        self.buffer_history = list()
-        for i in pbar:
-            # convert from a torch.Size object to a list
-            if self.verbose:
-                pbar.set_description("buffering images")
-            datum_index = shuffed_data_order[i]
-            datum = dataset[datum_index]
-            datum_size = datum[0].size()
-            self.buffer_history.append(datum_index)
-            self.buffer.append(datum)
-        self.remaining_to_be_in_buffer = shuffed_data_order[i + 1:]
-        assert len(patch_size) + 1 == len(datum_size)
-        self.patch_size = [datum_size[0]] + list(patch_size)
-
-    def __len__(self):
-        return self.npatches
-
-    def __getitem__(self, index):
+
+        self.counter = 0
+        self.epochs = -1  # incremented to 0 when buffer initially filled
+        self.buffer = deque()
+        self.remaining_to_be_in_buffer = deque()
+        self.buffer_history = []
+        pbar = tqdm(range(self.buffer_size))
+        pbar.set_description('Buffering images')
+        for _ in pbar:
+            self.insert_new_element_into_buffer()
+
+    def __iter__(self):
```

Comment:
Does this just return the same object on iter? Shouldn't this be like `super().__iter__()` or something?

Reply:
`BufferedPatchDataset` does not have a parent class so returning `super()`-something is not really relevant. The point of implementing `__iter__` is so users can do something like:

```python
bpds = BufferedPatchDataset(**some_kwargs)
for patch in bpds:
    process(patch)
```
Right now this is an infinite loop but might change in the future. e.g., stopping iteration after a specific number of patches or after a specific number of epochs through the input dataset.

Reply:
Ahhhh got it. Why not just make this your favourite thing then? `NotImplementedError` 🙃

Reply:
Well, because right now it is implemented. It's just an infinite loop, which might be just fine for some people if their use their own stopping criteria. :P